lua/luasnip/util/environ.lua

```diff
@@ -0,0 +1,176 @@
+local util = require("luasnip.util.util")
+
+local eager_vars = {
+	["TM_CURRENT_LINE"] = true,
+	["TM_CURRENT_WORD"] = true,
+	["TM_LINE_INDEX"] = true,
+	["TM_LINE_NUMBER"] = true,
+	["TM_SELECTED_TEXT"] = true,
+	["SELECT_RAW"] = true,
+	["SELECT_DEDENT"] = true,
+}
+-- These are the vars that have to be populated once the snippet starts to avoid any issue
+local function _fill_eager_vars(env)
+	local cur = util.get_cursor_0ind()
+	env.TM_CURRENT_LINE = vim.api.nvim_buf_get_lines(
+		0,
+		cur[1],
+		cur[1] + 1,
+		false
+	)[1]
+	env.TM_CURRENT_WORD = util.word_under_cursor(cur, env.TM_CURRENT_LINE)
+	env.TM_LINE_INDEX = tostring(cur[1])
+	env.TM_LINE_NUMBER = tostring(cur[1] + 1)
+	local selections = util.get_selection()
+	env.SELECT_RAW = selections[1]
+	env.SELECT_DEDENT = selections[2]
+	env.TM_SELECTED_TEXT = selections[3]
+end
+
+local lazy_vars = {}
+
+local Environ = {}
+function Environ:new(o)
+	o = o or {}
+	setmetatable(o, self)
+	_fill_eager_vars(o)
+	return o
+end
+
+function Environ:__index(key)
+	local v = lazy_vars[key]()
+	rawset(self, key, v)
+	return v
+end
+
+-- Variables defined in https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables
+
+-- Inherited from TextMate
+function lazy_vars.TM_FILENAME_BASE()
+	return vim.fn.expand("%:t:s?\\.[^\\.]\\+$??")
+end
+
+function lazy_vars.TM_DIRECTORY()
+	return vim.fn.expand("%:p:h")
+end
+function lazy_vars.TM_FILEPATH()
+	return vim.fn.expand("%:p")
+end
+
+-- Vscode only
+
+function lazy_vars.CLIPBOARD() -- The contents of your clipboard
+	return vim.fn.getreg('"', 1, true)
+end
+
+local function buf_to_ws_part()
+	local LSP_WORSKPACE_PARTS = "LSP_WORSKPACE_PARTS"
+	local ok, ws_parts = pcall(vim.api.nvim_buf_get_var, 0, LSP_WORSKPACE_PARTS)
+	if not ok then
+		local file_path = vim.fn.expand("%:p")
+
+		for _, ws in pairs(vim.lsp.buf.list_workspace_folders()) do
+			if file_path:find(ws, 1, true) == 1 then
+				ws_parts = { ws, file_path:sub(#ws + 2, -1) }
+				break
+			end
+		end
+		-- If it can't be extracted from lsp, then we use the file path
+		if not ok and not ws_parts then
+			ws_parts = { vim.fn.expand("%:p:h"), vim.fn.expand("%:p:t") }
+		end
+		vim.api.nvim_buf_set_var(0, LSP_WORSKPACE_PARTS, ws_parts)
```

Comment:
I think the current directory would make more sense here.

Reply:
the file can be out of the `cwd`

Reply:
I don't think that's too bad, wouldn't it be possible to get the relative path using `fnamemodify`?

Reply:
Yes, we can use `fnamemodify` but the file is supposed to be inside of the workspace path as far as I get from the vscode docs, which does not have to be the case with `cwd`.

Reply:
Oh, I didn't look at the docs, if there are some requirements it's alright if we keep it compatible.